tiledb/sm/filesystem/s3.cc

```diff
@@ -1332,37 +1345,66 @@ Status S3::write_multipart(
   const std::string uri_path(aws_uri.GetPath().c_str());
 
   // Take a lock protecting the shared multipart data structures
-  std::unique_lock<std::mutex> multipart_lck(multipart_upload_mtx_);
+  // Read lock to see if it exists
+  multipart_upload_mtx_.read_lock();
```

Comment:
I think it would be appropriate to grab the write lock here. I don't think we gain anything by using a read lock and then retrying with a write lock.

Reply:
Making an assumption, I believe the write to an existing part will be the most common use code path in this function. If that is true, then grabbing a write lock here will add a source of contention to the entire class. I think the overhead of switching locks is likely to be less than the exclusive write lock across this entire section.